pkgs/data/fonts/noto-fonts/default.nix

```diff
@@ -11,32 +11,48 @@
 , imagemagick
 , zopfli
 , buildPackages
+, fontconfig
+, python3
 }:
 
 let
-  mkNoto = { pname, weights }:
+  noto-pkg =
     stdenvNoCC.mkDerivation {
-      inherit pname;
-      version = "2020-01-23";
+      pname = "noto-fonts";
+      version = "unstable-2022-04-25";
 
       src = fetchFromGitHub {
         owner = "googlefonts";
         repo = "noto-fonts";
-        rev = "f4726a2ec36169abd02a6d8abe67c8ff0236f6d8";
-        sha256 = "0zc1r7zph62qmvzxqfflsprazjf6x1qnwc2ma27kyzh6v36gaykw";
+        rev = "a19de47f845dbd4c61b884c7ff90ce993555d05d";
+        sparseCheckout = ''
+          unhinted/otf
+          unhinted/variable-ttf
+        '';
+        hash = "sha256-7Jzo7402CiOcn2q++zbAzn2tP/y2x7QJuwjcP1NoyLQ=";
       };
 
+      analyzeScan = ./analyze_scan.py;
+
+      nativeBuildInputs = [ buildPackages.fontconfig.bin buildPackages.python3 ];
+
+      outputs = [ "out" "extra" "croscore" ];
+
       installPhase = ''
-        # We copy in reverse preference order -- unhinted first, then
-        # hinted -- to get the "best" version of each font while
-        # maintaining maximum coverage.
-        #
-        # TODO: install OpenType, variable versions?
-        local out_ttf=$out/share/fonts/truetype/noto
-        install -m444 -Dt $out_ttf phaseIII_only/unhinted/ttf/*/*-${weights}.ttf
-        install -m444 -Dt $out_ttf phaseIII_only/hinted/ttf/*/*-${weights}.ttf
-        install -m444 -Dt $out_ttf unhinted/*/*-${weights}.ttf
-        install -m444 -Dt $out_ttf hinted/*/*-${weights}.ttf
+        fc-scan ./unhinted -f '"%{fullname[0]}","%{file}"\n' > scan.csv
+        python $analyzeScan
+        local out_ttf=$out/share/fonts/google-noto
```

Comment:
Fonts are no longer installed to the `truetype` subdirectory. This seems like it would break at least `koreader`.

Reply:
Freetype scans all subdirectories regardless of name so the directory names are arbitrary. koreader [uses Freetype](https://github.com/koreader/crengine/blob/master/crengine/src/lvfntman.cpp#L5011) so should work fine. Certainly Fedora doesn't bother with the truetype directory anymore, see e.g. [Droid sans](https://koji.fedoraproject.org/koji/rpminfo?rpmID=30381402). Also it would be misleading in this case to call them "truetype" since they are a mixture of opentype (otf) and variable TTFs.

The more questionable aspect is whether koreader supports variable fonts. It seems to have the individual TTF file names names [hard-coded](https://github.com/koreader/koreader/blob/master/frontend/ui/font.lua#L39).  When I run koreader it crashes with the message ` frontend/ui/font.lua:353: attempt to index local 'face' (a nil value)`. This could be because it doesn't like the variable fonts I have installed.

Reply:
Sorry, my comment was more (or less?) concrete than that: The `koreader` derivation assumes this at the moment which should be easy to fix:

https://github.com/NixOS/nixpkgs/blob/fbdc30b994138a9e82fdb08b21781ab45fd81ef3/pkgs/applications/misc/koreader/default.nix#L46-L48

Reply:
Ah, I see. koreader should probably fetch the noto fonts itself in a private derivation, like [OnlyOffice](https://github.com/NixOS/nixpkgs/blob/634141959076a8ab69ca2cca0f266852256d79ee/pkgs/applications/office/onlyoffice-bin/default.nix#L57). Particularly it needs only a few languages and font variants, as compared to this package. cc @contrun and @neonfuz who can write that patch. (koreader crashes for me so I don't think I can do it)

Reply:
@a-m-joseph Any chance you could look into this, seeing as you recently touched `koreader`?

Reply:
> The more questionable aspect is whether koreader supports variable fonts. It seems to have the individual TTF file names names [hard-coded](https://github.com/koreader/koreader/blob/master/frontend/ui/font.lua#L39). When I run koreader it crashes with the message ` frontend/ui/font.lua:353: attempt to index local 'face' (a nil value)`. This could be because it doesn't like the variable fonts I have installed.

No, that means it couldn't find the fonts.

I get the same error above with this PR.

If I revert this PR I don't get the error.

> Ah, I see. koreader should probably fetch the noto fonts itself in a private derivation

Yeah, I'm not excited about this but I don't see any other way considering the hardcoded font names in koreader.  That project already has an astonishing level of vendoring going on.  I will write the PR to remove the `koreader`->`noto-fonts` dependency and have `koreader` do its own separate `fetchurl` to get the fonts.  I'm leaving town tomorrow morning and if I try to get it done before then I will end up doing a sloppy/rushed job.  This will need to wait until mid/late next week.  Feel free to ping me or open a bug and assign it to me as a reminder.

Reply:
> Unfortunately just convincing their CMake machinery to use pre-downloaded tarballs (instead of insisting on downloading them itself) has been so difficult

I've finally made some real progress on this.  Hope to have a draft PR shortly.

Reply:
> This seems like it would break at least `koreader`.

TL;DR: this problem disappears when we [we build koreader from source](https://github.com/NixOS/nixpkgs/pull/178320) instead of trying to rebundle the Debian binaries.

It turns out that nixpkgs' `koreader/default.nix` having a dependency upon nixpkgs' `noto-fonts` is 100% an artifact of the fact that `koreader/default.nix` is not much more than `dpkg -x`.  The problem disappears as soon as we build koreader from source.

Koreader has a [git submodule](https://github.com/koreader/koreader/tree/master/resources) containing the exact font files they expect (since they hardwire so many things, like the fonts' filenames).

These fonts are, for some reason, *not* included in the `.deb` archives that our current `koreader/default.nix` is using.  I suppose this could be in anticipation of Debian policy, although koreader has never been carried by Debian and would need a lot of other more drastic changes in order for them to accept it.  So I really don't know why the fonts get dropped when the `deb` is built.

In any event: my built-from-source `koreader` solves the problem here.